consensus/src/application/epoched.rs

```diff
@@ -0,0 +1,400 @@
+            unreachable!("missing starting epoch block at height {}", height);
+        };
+        block.commitment()
+    }
+
+    /// Proposes a new block or re-proposes the epoch boundary block.
+    ///
+    /// This method builds a new block from the underlying application unless the parent block
+    /// is the last block in the current epoch. When at an epoch boundary, it re-proposes the
+    /// boundary block to avoid creating blocks that would be invalidated by the epoch transition.
+    ///
+    /// The proposal operation is spawned in a background task and returns a receiver that will
+    /// contain the proposed block's commitment when ready. The built block is cached for later
+    /// broadcasting.
+    async fn propose(
+        &mut self,
+        context: Context<Self::Digest, S::PublicKey>,
+    ) -> oneshot::Receiver<Self::Digest> {
+        let mut marshal = self.marshal.clone();
+        let mut application = self.application.clone();
+        let last_built = self.last_built.clone();
+        let epoch_length = self.epoch_length;
+
+        // Metrics
+        let build_duration = self.build_duration.clone();
+
+        let (tx, rx) = oneshot::channel();
+        self.context
+            .with_label("propose")
+            .spawn(move |r_ctx| async move {
+                let (parent_view, parent_commitment) = context.parent;
+                let Ok(parent) = fetch_parent(
+                    parent_commitment,
+                    Some(Round::new(context.epoch(), parent_view)),
+                    &mut application,
+                    &mut marshal,
+                )
+                .await
+                else {
+                    warn!(
+                        ?parent_commitment,
+                        reason = "missing parent block",
+                        "skipping proposal"
+                    );
+                    return;
+                };
+
+                // Special case: If the parent block is the last block in the epoch,
+                // re-propose it as to not produce any blocks that will be cut out
+                // by the epoch transition.
+                let last_in_epoch = utils::last_block_in_epoch(epoch_length, context.epoch());
+                if parent.height() == last_in_epoch {
+                    let digest = parent.commitment();
+                    {
+                        let mut lock = last_built.lock().await;
+                        *lock = Some((context.round, parent));
+                    }
+
+                    let result = tx.send(digest);
+                    debug!(
+                        round = ?context.round,
+                        ?digest,
+                        success = result.is_ok(),
+                        "re-proposed parent block at epoch boundary"
+                    );
+                    return;
+                }
+
+                let start = Instant::now();
+                let Some(built_block) = application
```

Comment:
We should cancel this if `tx` is dropped via `.closed()`.

Reply:
We could use this as justification to remove the `timeout()` you wrap around the dkg but I know you felt strongly about that...